### PR TITLE
[Arith][GPU]Rewrite simplify fix for Vectorized Cooperative Fetching

### PR DESCRIPTION
**Pull request by jcf94**
This pr is part of #5883 , fix for the rewrite_simplify error when doing vectorized cooperative fetching in some cases.

Code generated with bug is shown like this:
```
A.shared[ramp(((ax0.ax1.fused.outer.outer*256) + (threadIdx.x_1*4)), 1, 4)] =
(float32x4*)A_2[(((broadcast(((floordiv(blockIdx.x, 4)*32768) + (ax0.ax1.fused.outer.outer*2048)), 4) + (floordiv(ramp((threadIdx.x_1*4), 1, 4), broadcast(64, 4))*broadcast(512, 4))) + broadcast((k.outer.outer*64), 4)) + floormod(ramp((threadIdx.x_1*4), 1, 4), broadcast(64, 4)))])
```
Which will finally lower to wrong CUDA C instructions.
This should be simplified to generate the correct RampNode:
```
A.shared[ramp(((ax0.ax1.fused.outer.outer*256) + (threadIdx.x_1*4)), 1, 4)] =
(float32x4*)A_2[ramp((((((floordiv(blockIdx.x, 4)*32768) + (ax0.ax1.fused.outer.outer*2048)) + (floordiv(threadIdx.x_1, 16)*512)) + (k.outer.outer*64)) + (floormod(threadIdx.x_1, 16)*4)), 1, 4)])
```

Then main problems inside this expression are:
```
threadIdx.x_1 = [0, 64]
floordiv(ramp((threadIdx.x_1*4), 1, 4), broadcast(64, 4)) * broadcast(512, 4)
floormod(ramp((threadIdx.x_1*4), 1, 4), broadcast(64, 4))
```
should be simplified to:
```
threadIdx.x_1 = [0, 64]
broadcast(floordiv(threadIdx.x_1, 16)*512), 4)
ramp(floormod(threadIdx.x_1, 16)*4, 1, 4)
```

@merrymercy Our former simplify rules will cause some extra bug, I find a better way to fix this.
@tqchen For the UTs, I'm not sure if there's any better way to check if all of the inner AST blocks are RampNode, the current UTs I added can still pass even if the vectorize failed.

cc @minminsun @FrozenGene @yangjunpro

### CODE REVIEW

**Comment by tqchen** [comment]
@jcf94 we should add a testcase to test_arith_rewrite_simplify, by constructing the case and verify the rules you have worked. Use `isinstance(x, tvm.ir.Ramp)` to assert the ramp node

File: src/arith/rewrite_simplify.cc [hunk 1]
```diff
@@ -722,8 +728,15 @@ PrimExpr RewriteSimplifier::Impl::VisitExpr_(const FloorDivNode* op) {
       ModularSet bmod = analyzer_->modular_set(b1.Eval());
       int64_t ramp_min = floordiv(bmod->base, c2val);
       int64_t ramp_max = floordiv(bmod->base + (lanes.Eval() - 1) * c1val, c2val);
-      if (bmod->coeff % c2val == 0 && ramp_min == ramp_max) {
-        return broadcast(floordiv(b1, c2), lanes).Eval();
+      if (ramp_min == ramp_max) {
+        // If b1 can device c2
```
Comment (tqchen) [inline src/arith/rewrite_simplify.cc:732]: device-> divide

File: src/arith/rewrite_simplify.cc [hunk 2]
```diff
@@ -722,8 +728,15 @@ PrimExpr RewriteSimplifier::Impl::VisitExpr_(const FloorDivNode* op) {
...
+        if (bmod->coeff % c2val == 0) {
+          return broadcast(floordiv(b1, c2), lanes).Eval();
+        }
+        // If all indices can be guaranteed to settle inside a coeff range
+        if (c2val % bmod->coeff == 0 && bmod->base + (lanes.Eval() - 1) * c1val < bmod->coeff) {
```
Comment (tqchen) [inline src/arith/rewrite_simplify.cc:737]: Please add a unit test in tests_arith_rewrite_simplify to cover this rule.

**Comment by jcf94** [comment]
> Thanks @jcf94 we should add a testcase to test_arith_rewrite_simplify, by constructing the case and
> 
> * verify each of the rule added in this PR.
> * Use `isinstance(x, tvm.ir.Ramp)` to assert the ramp node
> * You mentioned a bug in the previous rule, it would be great if the testcase covers the bug you mentioned

Commemts are all addressed.

- Add several test cases in tests_arith_rewrite_simplify for simplify rules
- Update test_target_codegen_cuda UTs & use pre post function to check the RampNode patterns
- The bug metioned above is introduced by our former implementation, currently everything work fine

**Comment by merrymercy** [comment]
@jcf94  Did our old rule affect the correctness of common operators?

**Comment by jcf94** [comment]
> @jcf94 Did our old rule affect the correctness of common operators?

Yes, with those rules several other UTs will fail, they're actually not always correct.
For example we have:
```
    TVM_TRY_REWRITE_IF(floordiv(x * c1 + y, c2), floordiv(x * c1, c2),
                       c1.Eval()->value > 0 && c2.Eval()->value > 0 &&
                       c2.Eval()->value % c1.Eval()->value == 0 &&
                       CanProveGreaterEqual(-y.Eval(), -c1.Eval()->value + 1));
```
while `floordiv(x * 4 + 4, 8)` cannot be simplified to `floordiv(x * 4, 8)`.

**Comment by tqchen** [comment]
Thanks @jcf94 @merrymercy . this PR is now merged

**Pull request merged**